### DISCUSSION TITLE
What kind of input can the Entity node handle?

### DISCUSSION THREAD

**Discussion by Timmiej93** [zachowj/node-red-contrib-home-assistant-websocket#557] [Q&A]
I have a JSON object like this:
```
{
  name: irrelevantData,
  name2: irrelevantData,
}
```

I want to create a `binary_sensor` for each name. They all have the same characteristics, only the name and the state changes (state is handles outside of JSON object). 

The `entity` node has an option for "input override". I tried messing with this, but the documentation tells me nothing about what kind of input this node can handle / expects. Is it possible to set the Home Assitant Config section (not the Attribute section) through input data?

I know I _could_ use a `split` and `switch` node to send each name to its respective `entity` node, but that sucks for reusability, since I'd have to add a `entity` node and change the `switch` node each time a new name is added to my JSON object.

**Comment by zachowj** [comment]
Valid inputs are listed <https://zachowj.github.io/node-red-contrib-home-assistant-websocket/node/entity.html#sensors-inputs>.

Not being able to set HA options of the entity dynamic is a limitation of the node. It needs those values to be able to register itself with HA at the time of deployment.

You can create entities dynamically using the API node and WebSocket endpoints the entity nodes uses. You can follow a past conversation on the topic [here](https://community.home-assistant.io/t/custom-component-for-node-red-contrib-home-assistant-websocket/150736/7?u=kermit).

**Reply by Timmiej93** [reply]
> Valid inputs are listed https://zachowj.github.io/node-red-contrib-home-assistant-websocket/node/entity.html#sensors-inputs.

Well don't I feel stupid now... I've seen that page at least a dozen times, each time thinking that only the very top section was about the entity node, and the rest being about other nodes, like the "configuration" node... 

> You can create entities dynamically using the API node and WebSocket endpoints the entity nodes uses. You can follow a past conversation on the topic [here](https://community.home-assistant.io/t/custom-component-for-node-red-contrib-home-assistant-websocket/150736/7?u=kermit).

I like the suggestion, but I'm having some trouble with getting it to work properly. The documentation is a bit lacking. Some questions:
- When do I need to send a command to `nodered/discovery`? Only once, after every reboot of HA? 
- What are `server_id` and `node_id`? Is `node_id` basically the text that comes after `sensor.`/`binary_sensor.`? `server_id` is a complete mystery to me. 
- Where can I find what kind of data Home Assistant requires? I can set the device class and icon easily, but I'm having a lot of trouble distinguishing between the `friendlyName` and `entityId`. Is there a way to set those two independently? If so, how?

**Reply by zachowj** [reply]
> The documentation is a bit lacking.

There is no documentation. It's not really meant for public user consumption. It could break it at any time.

> * When do I need to send a command to `nodered/discovery`? Only once, after every reboot of HA?

Yes, once after every HA reboot.

> * What are `server_id` and `node_id`? Is `node_id` basically the text that comes after `sensor.`/`binary_sensor.`? `server_id` is a complete mystery to me.

Home Assistant needs a unique identifier for each entity. The NR integration uses a form of the `server_id` and `node_id` at that identifier.

https://github.com/zachowj/hass-node-red/blob/2a11f2e0d96fac862a30be7c8b273f1b1acf53ea/custom_components/nodered/__init__.py#L112

The values can really be anything as long as they stay the same for each entity every time it is registered with HA.

> * but I'm having a lot of trouble distinguishing between the `friendlyName` and `entityId`. Is there a way to set those two independently? If so, how?

The friendly name is set by the property `config.name` during discovery. When HA sees a new unique id it will create the entity id based on the `name` property or default to something like `<domain>.nodered_<serverid>_<nodeid>`. After that first discovery of a new id changing the name property won't change the entity id only the friendly name. You can only change the entity id from within the HA UI or using other HA WebSocket endpoints.

**Reply by Timmiej93** [reply]
Thanks for the extensive reply, really helpful.

> Home Assistant needs a unique identifier for each entity. The NR integration uses a form of the `server_id` and `node_id` at that identifier.

Why are there two unique identifiers though? Can they both be the same string?

> After that first discovery of a new id changing the name property won't change the entity id only the friendly name.

What does "first discovery" mean in this instance, only the very first discovery, or every "first" discovery after HA reboot?

**Reply by zachowj** [reply]
> Why are there two unique identifiers though? Can they both be the same string?

In NR these are based on the node id of the entity node and node id of the server config. If you're running two instances of NR that happen to have an entity with the same node id in each instance, you imported one to the other or just random luck, the server id will still distinguish them from each other.

Yes, they can be the same string as long combined id is unique to each entity sent to HA.
 
> What does "first discovery" mean in this instance, only the very first discovery, or every "first" discovery after HA reboot?

Only the first time the discovery endpoint is hit not the first time after each reboot. You can do multiple discovery calls at any point to change any of the HA config options.

**Reply by Timmiej93** [reply]
Awesome, thank you very much for your help, this should be enough to get me going.
